### PR TITLE
Support for changing Docker Volume Basedir

### PR DESCRIPTION
**Pull request by shlee322**
When using the Remote Docker, there is a problem that the function code can not be found.

Modified to allow the host path to the volume to be specified, assuming that the specific path of the development machine is mounted on the Docker Host. (Ex: Docker Toolbox)

```
sam local invoke "TestFunc" -v /c/Users/shlee/projects/test-project
```

### CODE REVIEW

**Comment by sanathkr** [comment]
The base directory is usually SAM file's directory. You could always set CodeUri to point to your function code relative to SAM file, right? 

1. Why would you need this change? 
2. Can we prefix the command flat with --docker so we know it's clearly a docker flag? I would use the same flag name that docker CLI uses

**Comment by shlee322** [comment]
@sanathkr 

1. There is no problem when using the locally installed Docker, but a problem occurs if the Docker Host is remote machine.
(The problem also occurs when using the Docker Toolbox.)

```
Mount (ex: NFS)
Docker Host: /mnt/projects
Developer Machine: /home/shlee322/projects

[Developer Machine Cmd]

> export DOCKER_HOST=tcp://192.168.99.100:2376
> cd /home/shlee322/projects/test
> sam local invoke "testfunc"

Unable to import module 'lambda_function': No module named 'lambda_function'

# docker -v /home/shlee322/projects/test/testfunc:/var/task:ro
# The path does not exist.

sam local invoke "testfunc" -v /mnt/projects/test

# docker -v /mnt/projects/test/testfunc:/var/task:ro
# success!
```

2. Thanks for your advice. I will change it.

**Comment by sanathkr** [comment]
Now this makes sense. Can you update the readme along with your commit? Just add a section under Advanced Usages to connect to remote docker host?

**Comment by sanathkr** [comment]
alsl, for sake of my understanding, do you have a nfs mount that shares your file system with a remote docker host? How does the mount work at first place when docker is running separate machine?

**Comment by shlee322** [comment]
@sanathkr 

I will commit to reflect your comments.

To use NFS, the development computer must be mounted on the remote host in advance.
I think there are a lot of ways to connect a development computer to a remote host, so it's hard to explain everything.

For the Readme, I will use the Docker Toolbox as an example.
When you install and run Docker Toolbox, the Linux VM with Docker is automatically installed in the virtual box.
The /c/ path for this Linux VM is automatically shared with C:\ on the host machine.

ps. 

I'm using bash to install the Windows 10 linux subsystem on this Windows machine,
For this subsystem, /mnt/c/ is C:\ on the host machine.

I found this issue using bash with aws-sam-local.
(path : /mnt/c/Users/shlee322/projects/testproject)

**Comment by shlee322** [comment]
@sanathkr 
I have added some modify code and README.md.
Please your review.

File: runtime.go [hunk 1]
```diff
@@ -573,12 +573,8 @@ func getWorkingDir(basedir string, codeuri string) (string, error) {
 
 	dir := filepath.Join(abs, codeuri)
 
-	// ...but only if it actually exists
```
Comment (sanathkr) [inline runtime.go]: Why was this removed?

File: main.go [hunk 2]
```diff
@@ -117,6 +122,11 @@ func main() {
 							Usage: "Optional. When specified, Lambda function container will start in debug mode and will expose this port on localhost.",
 							EnvVar: "SAM_DEBUG_PORT",
 						},
+						cli.StringFlag{
+							Name:   "docker-volume-basedir, v",
+							Usage:  "Optional. Docker Volume Host Path",
```
Comment (sanathkr) [inline main.go]: Can you write a more detailed Usage explanation. This is a bit too less

File: README.md [hunk 3]
```diff
@@ -371,6 +372,22 @@ Example:
 $ sam local invoke --log-file ./output.log
 ```
 
+
+### Remote Docker
+Sam Local loads the function code using Docker Volume. As a result, The project directory must be pre-mounted on the remote host where the Docker is running.
```
Comment (sanathkr) [inline README.md]: I would say "loads function code by _mounting_ filesystem to a Docker Volume". Just to make it obvious.

File: runtime.go (hunk 1 above)
Comment (shlee322) [inline runtime.go]: If using docker-volume-basedir, because basedir is the path of the remote system, you will not know if basedir exists using os.Stat on the machine running aws-sam-local.

File: README.md (hunk 3 above)
Comment (shlee322) [inline README.md]: I will commit to reflect your comments.

File: main.go (hunk 2 above)
Comment (shlee322) [inline main.go]: I will commit to reflect your comments.

**Comment by shlee322** [comment]
@sanathkr I modified the description for docker-volume-basedir.
My English skills are not enough and there may be grammatical problems.
Please your review. Thank  :)

File: runtime.go (hunk 1 above)
Comment (sanathkr) [inline runtime.go]: That's fine, if os.stat fails, we will default to basedir which is the mount point passed thru CLI paramerter right? 

I think we should not remove this code. What you think?

Comment (shlee322) [inline runtime.go]: Added NewRuntimeOpt.CheckWorkingDirExist option. Thank you.

**Comment by sanathkr** [comment]
Merged! Thanks for making SAM Local CLI awesome!

**Comment by shlee322** [comment]
Thanks! :)
